### DISCUSSION TITLE
Find One on view=> curl: (52) Empty reply from server

### DISCUSSION THREAD

**Discussion by hamiltonsalazar** [apache/couchdb#2932] [Q&A] [answered]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

I followed the instructions in the documentation for [Find One](https://docs.couchdb.org/en/stable/ddocs/views/intro.html#find-one) on a view, but that don't work for me.

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Current Behavior

I followed every step to create a view with the function map: 

`function(doc) {
    if(doc.date && doc.title) {
        emit(doc.date, doc.title);
    }
}`

And i have aggregate the three example documents. 

When i query the view with `/database/_design/designdocname/_view/viewname` for me `curl -X GET http://anna:secret@localhost:5984/views/_design/myDesignDoc/_view/firstView` I get the next result:

> `{
    "total_rows":3,"offset":0,"rows":[
        {
            "id":"hello-world",
            "key":"2009/01/15 15:52:20",
            "value":"Hello World"
        },
        {
            "id":"biking",
            "key":"2009/01/30 18:04:11",
            "value":"Biking"
        },
        {
            "id":"bought-a-cat",
            "key":"2009/02/17 21:13:39",
            "value":"Bought a Cat"
        }
    ]
}`

But when I try to find a single document with `/blog/_design/docs/_view/by_date?key="2009/01/30 18:04:11"` doesn't work. 

I use `curl -X GET http://anna:secret@localhost:5984/views/_design/myDesignDoc/_view/by_date?key="2009/01/30 18:04:11"` and CouchDBrespond with: 

> curl: (52) Empty reply from server

Also i try with `curl -X GET http://anna:secret@localhost:5984/views/_design/myDesignDoc/_view/firstView/by_date?key="2009/01/30 18:04:11"` and CouchDBrespond again with: 

> curl: (52) Empty reply from server

[NOTE]: # ( Include commands to reproduce, if possible. curl is preferred. )

## Expected Behaviour

I should get the “Biking” blog post: 

> {
            "id":"biking",
            "key":"2009/01/30 18:04:11",
            "value":"Biking"
        }

[NOTE]: # ( Tell us what you expected to happen. )

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB version used: 3.1.0
* Operating system and version: Windows 10 Pro 1909

## Additional Context

I have the three documents load in CouchDB: 
![image](https://user-images.githubusercontent.com/36459473/83995361-d1ab5100-a91e-11ea-8910-8a10f99e69b6.png)

[TIP]:  # ( Add any other context about the problem here. )

**Comment by wohali** [comment, marked as answer]
This is probably escaping biting you.

Here's the full example on 3.1.0 driven from Linux. In this example, my CouchDB is at `http://localhost:5984/`.

```bash
$ curl -X PUT http://localhost:15984/abc
{"ok":true}

$ curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"biking","title":"Biking","body":"My biggest hobby is mountainbiking. The other day...","date":"2009/01/30 18:04:11"}'
{"ok":true,"id":"biking","rev":"1-9e21110dc665c8e0201643aa3ad3aeaf"}

$ curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"bought-a-cat","title":"Bought a Cat","body":"I went to the the pet store earlier and brought home a little kitty...","date":"2009/02/17 21:13:39"}'
{"ok":true,"id":"bought-a-cat","rev":"1-ccd0a7e95ade7eb2d29cc02ffa582b30"}

$ curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"hello-world","title":"Hello World","body":"Well hello and welcome to my new blog...","date":"2009/01/15 15:52:20"}'
{"ok":true,"id":"hello-world","rev":"1-97dd85b06c25328a300f3f4041def370"}

$ curl http://localhost:15984/abc/_all_docs
{"total_rows":3,"offset":0,"rows":[
{"id":"biking","key":"biking","value":{"rev":"1-9e21110dc665c8e0201643aa3ad3aeaf"}},
{"id":"bought-a-cat","key":"bought-a-cat","value":{"rev":"1-ccd0a7e95ade7eb2d29cc02ffa582b30"}},
{"id":"hello-world","key":"hello-world","value":{"rev":"1-97dd85b06c25328a300f3f4041def370"}}
]}

$ curl -X PUT -H "Content-Type:application/json" http://localhost:15984/abc/_design/docs -d '{"language":"javascript","views": {"by_date":{"map":"function(doc) {if(doc.date && doc.title) { emit(doc.date, doc.title) } };"}}}'
{"ok":true,"id":"_design/docs","rev":"1-e961555ed4f1dc188a6610d092b68393"}

$ curl http://localhost:15984/abc/_design/docs/_view/by_date
{"total_rows":3,"offset":0,"rows":[
{"id":"hello-world","key":"2009/01/15 15:52:20","value":"Hello World"},
{"id":"biking","key":"2009/01/30 18:04:11","value":"Biking"},
{"id":"bought-a-cat","key":"2009/02/17 21:13:39","value":"Bought a Cat"}
]}

$ curl 'http://localhost:15984/abc/_design/docs/_view/by_date?key="2009/01/30%2018:04:11"'
{"total_rows":3,"offset":1,"rows":[
{"id":"biking","key":"2009/01/30 18:04:11","value":"Biking"}
]}
```

Note the use of `%20` to escape the space.

If you're running curl on Windows, you need to go through some extra machinations to get curl to correctly escape the content. You can't use single-quotes, only double quotes, which means the internal double-quotes around the key have to be escaped like this:

```
curl "http://localhost:15984/abc/_design/docs/_view/by_date?key=\"2009/01/30%2018:04:11\""
```

Hope this works for you - if so, please mark my answer as correct! :wink:

**Reply by hamiltonsalazar** [reply]
The command `curl "http://localhost:15984/abc/_design/docs/_view/by_date?key=\"2009/01/30%2018:04:11\""` worked for me, i tried this code with changes for Windows and almost all works, except when i wanted to create my view, i'll leave my code here to help other people who run curl on Windows. I used Project Fauxton for create the view, my CouchDB is at http://localhost:5984/ and i'm using the user and password of the documentation example 

```
curl -X PUT http://localhost:15984/abc 

curl -X PUT http://anna:secret@localhost:5984/abc

{"ok":true}



curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"biking","title":"Biking","body":"My biggest hobby is mountainbiking. The other day...","date":"2009/01/30 18:04:11"}'

curl -X POST -H "Content-Type:application/json" http://anna:secret@localhost:5984/abc -d "{\"_id\":\"biking\",\"title\":\"Biking\",\"body\":\"My biggest hobby is mountainbiking. The other day...\",\"date\":\"2009/01/30 18:04:11\"}"

{"ok":true,"id":"biking","rev":"1-9e21110dc665c8e0201643aa3ad3aeaf"}



curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"bought-a-cat","title":"Bought a Cat","body":"I went to the the pet store earlier and brought home a little kitty...","date":"2009/02/17 21:13:39"}'

curl -X POST -H "Content-Type:application/json" http://anna:secret@localhost:5984/abc -d "{\"_id\":\"bought-a-cat\",\"title\":\"Bought a Cat\",\"body\":\"I went to the the pet store earlier and brought home a little kitty...\",\"date\":\"2009/02/17 21:13:39\"}"

{"ok":true,"id":"bought-a-cat","rev":"1-ccd0a7e95ade7eb2d29cc02ffa582b30"}



curl -X POST -H "Content-Type:application/json" http://localhost:15984/abc -d '{"_id":"hello-world","title":"Hello World","body":"Well hello and welcome to my new blog...","date":"2009/01/15 15:52:20"}'

curl -X POST -H "Content-Type:application/json" http://anna:secret@localhost:5984/abc -d "{\"_id\":\"hello-world\",\"title\":\"Hello World\",\"body\":\"Well hello and welcome to my new blog...\",\"date\":\"2009/01/15 15:52:20\"}"

{"ok":true,"id":"hello-world","rev":"1-97dd85b06c25328a300f3f4041def370"}



curl http://localhost:15984/abc/_all_docs

curl http://anna:secret@localhost:5984/abc/_all_docs

{"total_rows":3,"offset":0,"rows":[
{"id":"biking","key":"biking","value":{"rev":"1-9e21110dc665c8e0201643aa3ad3aeaf"}},
{"id":"bought-a-cat","key":"bought-a-cat","value":{"rev":"1-ccd0a7e95ade7eb2d29cc02ffa582b30"}},
{"id":"hello-world","key":"hello-world","value":{"rev":"1-97dd85b06c25328a300f3f4041def370"}}
]}



curl -X PUT -H "Content-Type:application/json" http://localhost:15984/abc/_design/docs -d '{"language":"javascript","views": {"by_date":{"map":"function(doc) {if(doc.date && doc.title) { emit(doc.date, doc.title) } };"}}}'

curl -X PUT -H "Content-Type:application/json" http://anna:secret@localhost:5984/abc/_design/docs -d "{\"language\":\"javascript\",\"views\": {\"by_date\":{\"map\":\" function(doc) {if(doc.date && doc.title) { emit(doc.date, doc.title) } }; \"}}}"

{"error":"bad_request","reason":"invalid UTF-8 JSON"}
"doc.title)" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.



curl http://localhost:15984/abc/_design/docs/_view/by_date

curl http://anna:secret@localhost:5984/abc/_design/docs/_view/by_date

{"total_rows":3,"offset":0,"rows":[
{"id":"hello-world","key":"2009/01/15 15:52:20","value":"Hello World"},
{"id":"biking","key":"2009/01/30 18:04:11","value":"Biking"},
{"id":"bought-a-cat","key":"2009/02/17 21:13:39","value":"Bought a Cat"}
]}



curl 'http://localhost:15984/abc/_design/docs/_view/by_date?key="2009/01/30%2018:04:11"'

curl "http://anna:secret@localhost:5984/abc/_design/docs/_view/by_date?key=\"2009/01/30%2018:04:11\""

{"total_rows":3,"offset":1,"rows":[
{"id":"biking","key":"2009/01/30 18:04:11","value":"Biking"}
]}
```

@wohali maybe you can take a look this part: 

```
curl -X PUT -H "Content-Type:application/json" http://localhost:15984/abc/_design/docs -d '{"language":"javascript","views": {"by_date":{"map":"function(doc) {if(doc.date && doc.title) { emit(doc.date, doc.title) } };"}}}'

curl -X PUT -H "Content-Type:application/json" http://anna:secret@localhost:5984/abc/_design/docs -d "{\"language\":\"javascript\",\"views\": {\"by_date\":{\"map\":\" function(doc) {if(doc.date && doc.title) { emit(doc.date, doc.title) } }; \"}}}"

{"error":"bad_request","reason":"invalid UTF-8 JSON"}
"doc.title)" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.
```

**Reply by wohali** [reply]
`&&` is a Windows/DOS operator. You need to escape the ampersands.

**Reply by hamiltonsalazar** [reply]
@wohali Thank you very much! 

`^&^&`
